### ISSUE TITLE
Question, hiding loaderDialog cancels other events

### ISSUE THREAD

**Issue by jalanga** [felangel/bloc#770]
In a page I have 2 or more async requests, for every async request I open the loader dialog with `showDialog`, if is not opened, and after every request I try to close it with `Navigator.of(context, rootNavigator: true).pop();` if is opened.
I was using flutter_bloc: 0.20.1, I upgraded to the latest flutter_bloc, with the latest flutter version. Now I am flutter_bloc: 2.1.1, and flutter: 1.10.15-pre.307

In my bloc I do the following thing:

```
    if (event is LoadCheckoutInfo){
      yield CheckoutShowLoader();
      try{
        Map<String, dynamic> checkoutInfo;
        checkoutInfo = await UserRepository.getCheckoutInfo(event.planId, event.billingCycles, event.couponCode);
        yield LoadCheckoutInfoDone(checkoutInfo: checkoutInfo);
        yield CheckoutHideLoader();
      } catch(error){
        yield CheckoutHideLoader();
        yield LoadCheckoutError(error: error.response);
      }
    } else if(event is GetBillingInfo){
      yield CheckoutShowLoader();
      try{
        Map<String, dynamic> userBillingInfo;
        userBillingInfo = await UserRepository.getUserBillingInfo();
        yield LoadBillingInfoDone(userBillingInfo: userBillingInfo);
        yield CheckoutHideLoader();
      } catch(error){
        yield CheckoutHideLoader();
        yield CheckoutError(error: error.response["error"]);
      }
    }
```

The problem is when I call `LoadCheckoutInfoDone` I can catch the state on `BlocListener`, after that I receive on `BlocListener` the state `CheckoutHideLoader`, I hide the loader, but I will not receive `LoadCheckoutInfoDone` on my `BlocBuilder`. 
I saw that events are first entering on `BlocListener` and after that on `BlocBuilder`.
I have this problem with loaders on all my application.

How can I make `Navigator.of(context, rootNavigator: true).pop();` to work without, probably closing the stream?

Update:

Code for loader dialog:

```
void showLoadingDialog(BuildContext context) {
  showDialog(
      context: context,
      barrierDismissible: false,
      builder: (BuildContext context) {
        return Theme(
            data: Theme.of(context).copyWith(
                dialogBackgroundColor: Colors.transparent),
            child: new Dialog(
              child: new LoadingIndicator(),
            )
        );
      }
  );
}
```

***Update 2***
Seems that not showing/hiding the loader is the problem, the problem seems to be yield-ing 2 states one after another one, the states is catch just in `BlocListener` and not entering in `BlocBuilder`

Here is my `BlocBuilder`, I commented the lines of showing/hiding the loader to achieve this "error" :

```
@override
  Widget build(BuildContext context) {
    return BlocBuilder<CheckoutBloc, CheckoutState>(builder: (context, state) {
      if (state is LoadCheckoutInfoDone) {
        this.checkoutInfo = state.checkoutInfo;
      } else if (state is LoadUserCompaniesDone){
        userCompanies = state.userCompanies;
        if(userCompanies.length != 0 && _selectedCompany == null){
            _selectedCompany = userCompanies[0]['id'];
        }
        if (state.countryId != null && _isNewAddedCompany == true) {
          _selectedCompany = state.countryId;
        }
      }
      var addCompanyDialogInstance = AddCompanyDialog(
          context: context,
          onSubmit: _getCompanies,
          onCancel: () => (){},
          countryCode: _userBillingInfo['country']
      );

      var editBillingDialog = BillingInfoDialog(
        context: context,
        billingInfo: _userBillingInfo,
        onSubmit: _getBillingInfo,
        onCancel: () => (){},
      );
      return MultiBlocListener(
        listeners: [
          BlocListener<AuthenticationBloc, AuthenticationState>(
            listener: (context, state) {
              if (state is AuthenticationUnauthenticated) {
                Navigator.of(this.context).pushNamedAndRemoveUntil(
                    Routes.auth, (Route<dynamic> route) => false);
              }
            },
          ),
          BlocListener<CheckoutBloc, CheckoutState>(
            listener: (context, state) {
              if (state is PayNowComplete) {
                launchUrl(state.payNowUrl);
              } else if (state is CheckoutShowLoader) {
                if (_openedLoaders == 0) {
//                  showLoadingDialog(context);
                }
                _openedLoaders += 1;
              } else if (state is LoadCheckoutError) {
                SchedulerBinding.instance.addPostFrameCallback(
                        (_) => showFlush(context, "Error", state.error["error"]));
                if (state.error["data"] != null &&
                    state.error["data"]["validation_errors"] != null) {
                  SchedulerBinding.instance
                      .addPostFrameCallback((_) => setState(() {
                    _couponErrorText = state.error["data"]
                    ["validation_errors"]["coupon"][0];
                  }));
                }
              } else if (state is LoadUserCompaniesError){
                SchedulerBinding.instance.addPostFrameCallback(
                        (_) => showFlush(context, "Error", state.error));
              } else if (state is CheckoutError){
                SchedulerBinding.instance.addPostFrameCallback(
                        (_) => showFlush(context, "Error", state.error));
              } else if(state is LoadBillingInfoDone){
                _userBillingInfo = state.userBillingInfo;
              } else if (state is CheckoutHideLoader) {
                if (_openedLoaders == 1) {
//                  Navigator.of(context).pop('dialog');
                }
                _openedLoaders -=1;
              }

            },
          ),
        ],
        child: Scaffold(
```

Transitions:
```
I/flutter ( 6329): Transition { currentState: CheckoutInitial, event: LoadCheckoutInfo {planId: 1; billingCycles: 1; coupon: null}, nextState: CheckoutShowLoader }
I/flutter ( 6329): Transition { currentState: CheckoutShowLoader, event: LoadCheckoutInfo {planId: 1; billingCycles: 1; coupon: null}, nextState: LoadCheckoutInfoDone { locations: {plan_name: Degeaba,} } }
I/flutter ( 6329): Transition { currentState: LoadCheckoutInfoDone { locations: {plan_name: Degeaba} }, event: LoadCheckoutInfo {planId: 1; billingCycles: 1; coupon: null}, nextState: CheckoutHideLoader }
I/flutter ( 6329): Transition { currentState: CheckoutHideLoader, event: GetBillingInfo, nextState: CheckoutShowLoader }
I/flutter ( 6329): Transition { currentState: CheckoutShowLoader, event: GetBillingInfo, nextState: LoadBillingInfoDone { userCompanies: {} } }
I/flutter ( 6329): Transition { currentState: LoadBillingInfoDone { userCompanies: {card_holder_name: 1} }, event: GetBillingInfo, nextState: CheckoutHideLoader }
```

**Comment by rodrigobastosv** [comment]
Hi @jalanga do you mind showing your BlocBuilder / BlocListener code?

My guess is that you are yielding LoadCheckoutInfoDone state two times, and the way BlocBuilder works is that it don't execute it's build if the state yielded is the same as the previous state.

**Comment by jalanga** [comment]
Hi @rodrigobastosv 

I don't think I am yield-ing 2 time, I updated my question, and I will put the transitions.
Thank you

**Comment by rodrigobastosv** [comment]
> The problem is when I call LoadCheckoutInfoDone I can catch the state on BlocListener, after that I receive on BlocListener the state CheckoutHideLoader, I hide the loader, but I will not receive LoadCheckoutInfoDone on my BlocBuilder.

Hi again, i analised here. You are yielding the CheckoutHideLoader after LoadCheckoutInfoDone. I think you have to make this way:
```
yield CheckoutHideLoader();
yield LoadCheckoutInfoDone(checkoutInfo: checkoutInfo);
```

So you will garantee LoadCheckoutInfoDone will be yielded after CheckoutHideLoader

**Comment by jalanga** [comment]
@rodrigobastosv 
Is not working,  here is what I receive:

```
I/flutter (11975): Builder - CheckoutInitial
I/flutter (11975): Listener - CheckoutShowLoader
I/flutter (11975): Builder - CheckoutShowLoader
I/flutter (11975): Listener - CheckoutHideLoader
I/flutter (11975): Listener - LoadCheckoutInfoDone { }
I/flutter (11975): Listener - CheckoutShowLoader
I/flutter (11975): Builder - CheckoutShowLoader
I/flutter (11975): Listener - CheckoutHideLoader
I/flutter (11975): Listener - LoadBillingInfoDone {  }
I/flutter (11975): Builder - LoadBillingInfoDone { }
```
And I don't show/hide loaders.
I don't think that the order counts.
It seems that triggering 2 events one after another will break.

If I make just 1 event is kind of working:
HideLoader will not arrive in Builder is just in Listener
```
I/flutter (11975): Builder - CheckoutInitial
I/flutter (11975): Listener - CheckoutShowLoader
I/flutter (11975): Builder - CheckoutShowLoader
I/flutter (11975): Listener - CheckoutHideLoader
I/flutter (11975): Listener - LoadCheckoutInfoDone { }
I/flutter (11975): Builder - LoadCheckoutInfoDone { }
```

Is this a bug? I wanted to release the app on Monday :)

**Comment by felangel** [comment]
Hi @jalanga 👋 
Thanks for opening an issue!

Are you able to share a link to a sample app which illustrates the issue you're having? It would be much easier to help if I'm able to run/reproduce the problem locally. Thanks!

**Comment by jalanga** [comment]
@felangel 
Hi, 
I am sorry for delay.
I made an example over your example of Timer:
https://github.com/jalanga/FlutterBlocIssue

I just added 2 fake states:
https://github.com/jalanga/FlutterBlocIssue/blob/master/lib/bloc/timer_state.dart
And on reset I emit 3 states: FakeStateOne, Ready, FakeStateTwo:
https://github.com/jalanga/FlutterBlocIssue/blob/master/lib/bloc/timer_bloc.dart

Every state arrives in condition, every state passes the condition, but just FakeStateTwo arrives in builder.
In my code, I want to show loader, load data from server and hide loader. 
I am using BlocListener inside BlocBuilder, every state arrives in BlocListener but not in BlocBuilder.
I changed the order, using Builder inside Listener, same result.
